kmir/src/kmir/kdist/mir-semantics/kmir.k

```diff
@@ -14,167 +15,717 @@ module KMIR-SYNTAX
+        => #evalArgs( ARGS, ListItem( #evalMirConst( TY, CID, KIND ) ) EVALED )
            ...
        </k>
-       <callStack> Stack => ListItem(CallRecord( FnAddr( CallerAddr ), Target, Dest, Unwind, CallerLocals )) Stack </callStack>
-       <body> _CallerBody => CalleeBody </body>
-       <locals> CallerLocals => #makeLocals( LocalCount, ArgCount, Args ) </locals>
-       <funcAddr> FnAddr( CallerAddr => CalleeAddr ) </funcAddr>
 
-  rule <k> basicBlock( .Statements, terminator( terminatorKindReturn, _SPAN ))
-        => #goto( Target:MaybeBasicBlockIdx )
+  rule <k> #evalArgs( ListItem( PLACE:Place ) ARGS, EVALED )
+        => #evalArgs( ARGS, ListItem( #loadPlace( PLACE ) ) EVALED )
            ...
        </k>
-       <callStack> ListItem(CallRecord( FnAddr( CallerAddr ), Target, Dest, _Unwind, CallerLocals )) Stack => Stack </callStack>
-       <locals> (RetValue _Locals:Values) => #setValue( CallerLocals, Dest, RetValue ) </locals>
-       <body> _CalleeBody => CallerBody </body>
-       <memory> ... CallerAddr |-> FnVal( CallerBody, _, _ ) ... </memory>
-    requires Target =/=K noBasicBlockIdx
 
-  syntax KItem ::= panicked() [symbol(panicked)]
+  rule <k> #evalArgs( .List, ARGS ) => #operands( ARGS ) ... </k>
+
+  syntax Value ::= #evalMirConst( Ty, Int, ConstantKind ) [function]
+                 | #evalTyConst( TyConstKind )            [function]
+
+  rule #evalMirConst( _TY, _CID, constantKindZeroSized                                                     ) => #makeStructValue ( 0, .List )
+  rule #evalMirConst(  TY, _CID, constantKindAllocated( ALLOC )                                            ) => #loadAllocation( TY, ALLOC )
+  rule #evalMirConst(  TY, _CID, constantKindUnevaluated( unevaluatedConst( DEF, _ARGS, PROMOTED_INDEX ) ) ) => #loadUnevalConst( TY, DEF, PROMOTED_INDEX )
+  rule #evalMirConst( _TY, _CID, constantKindTy( tyConst( KIND, _ID ) )                                    ) => #evalTyConst( KIND )
+  // Handle extended function constant types
+  rule #evalMirConst( _TY, _CID, constantKindFnDef( Addr )                                                 ) => #makeFnPtrValue( Addr )
+  rule #evalMirConst( _TY, _CID, constantKindNoOp                                                          ) => #getNoOpFnAddr()
+  rule #evalMirConst( _TY, _CID, constantKindIntrinsic( Sym )                                              ) => #getIntrinsicAddr( Sym )
+
+  rule #evalTyConst( tyConstKindZSTValue( _TY )           ) => #makeStructValue( 0, .List )
+  rule #evalTyConst( tyConstKindValue( TY, ALLOC )        ) => #loadAllocation( TY, ALLOC )
+  // NOTE: tyConstKindUnevaluated( DEF, ARGS ) should not occur --- all type system constants must be fully evaluated
+
+  syntax Value ::= #getArg( List )     [function]
+                 | #getFstArg( List )  [function]
+  syntax List ::= #getRestArgs( List ) [function]
+
+  rule #getArg( ListItem( Val:Value ) )     => Val
+  rule #getFstArg( ListItem( Val:Value ) )  => Val
```

Comment:
`getArg` and `getFirstArg` are exactly the same, is this intentional? What is the use case of each? Is `getArg` for the single item of a list of one element and `getFirstArg` for the first item of a list of any number of elements?

Reply:
This is a bug. `getFirstArg` should get the head of a list --- which it currently doesn't do.